tsnet/packet_filter_test.go

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package tsnet
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"testing"
+	"time"
+
+	"tailscale.com/ipn"
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/ipproto"
+	"tailscale.com/types/key"
+	"tailscale.com/types/netmap"
+	"tailscale.com/util/must"
+	"tailscale.com/wgengine/filter"
+)
+
+// waitFor blocks until a NetMap is seen on the IPN bus that satisfies the given
+// function f. Note: has no timeout, should be called with a ctx that has an
+// appropriate timeout set.
+func waitFor(t testing.TB, ctx context.Context, s *Server, f func(*netmap.NetworkMap) bool) error {
+	t.Helper()
+	t.Log("starting waitFor")
+	watcher, err := s.localClient.WatchIPNBus(ctx, ipn.NotifyInitialNetMap)
+	if err != nil {
+		t.Fatalf("error watching IPN bus: %s", err)
+	}
+	defer watcher.Close()
+
+	for {
+		n, err := watcher.Next()
+		if err != nil {
+			return fmt.Errorf("getting next ipn.Notify from IPN bus: %w", err)
+		}
+		if n.NetMap != nil {
+			if f(n.NetMap) {
+				return nil
+			}
+		}
+	}
+}
+
+// TestPacketFilterFromNetmap tests all of the client code for processing
+// netmaps and turning them into packet filters together. Only the control-plane
+// side is mocked out.
+func TestPacketFilterFromNetmap(t *testing.T) {
+	t.Parallel()
+
+	var key key.NodePublic
+	must.Do(key.UnmarshalText([]byte("nodekey:5c8f86d5fc70d924e55f02446165a5dae8f822994ad26bcf4b08fd841f9bf261")))
+
+	type check struct {
+		src  string
+		dst  string
+		port uint16
+		want filter.Response
+	}
+
+	tests := []struct {
+		name            string
+		initialNetMap   *tailcfg.MapResponse
```

Comment:
initialMapResponse? initialRes? I've found referring to the bytes on the wire as MapResponse to be a good delineator